utils_cv/detection/model.py

```diff
@@ -285,63 +380,138 @@ def plot_precision_loss_curves(
         """ Plot training loss from calling `fit` and average precision on the
         test set. """
         fig = plt.figure(figsize=figsize)
-        ax1 = fig.add_subplot(111)
+        ap = {k: [dic[k] for dic in self.ap] for k in self.ap[0]}
 
-        ax1.set_xlim([0, self.epochs - 1])
-        ax1.set_xticks(range(0, self.epochs))
-        ax1.set_title("Loss and Average Precision over epochs")
-        ax1.set_xlabel("epochs")
-        ax1.set_ylabel("loss", color="g")
-        ax1.plot(self.losses, "g-")
+        for i, (k, v) in enumerate(ap.items()):
 
-        ax2 = ax1.twinx()
-        ax2.set_ylabel("average precision", color="b")
-        ax2.plot(self.ap, "b-")
+            ax1 = fig.add_subplot(1, len(ap), i+1)
 
-    def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
+            ax1.set_xlim([0, self.epochs - 1])
+            ax1.set_xticks(range(0, self.epochs))
+            ax1.set_title("Loss and Average Precision over epochs")
+            ax1.set_xlabel("epochs")
+            ax1.set_ylabel("loss", color="g")
+            ax1.plot(self.losses, "g-")
+
+            ax2 = ax1.twinx()
+            ax2.set_ylabel("average precision for ".format(k), color="b")
+            ax2.plot(v, "b-")
+
+    def evaluate(self, dl: DataLoader = None) -> Union[CocoEvaluator, None]:
         """ eval code on validation/test set and saves the evaluation results in self.results. """
         if dl is None:
+            if not self.dataset:
+                return
             dl = self.dataset.test_dl
         self.results = evaluate(self.model, dl, device=self.device)
         return self.results
 
+    def _transform(
+        self,
+        im: Union[str, Path, Image.Image, np.ndarray],
+    ) -> torch.Tensor:
+        """ Convert the image to the format required by the model. """
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im)
+        if self.device:
+            im = im.to(self.device)
+        return im
+
+    @classmethod
+    def _apply_threshold(
+        cls,
+        pred: Dict,
+        threshold: Optional[int] = 0.5,
+        mask_threshold: Optional[int] = 0.5,
+        **kwargs,
+    ) -> Dict:
+        """ Return prediction results that are above the threshold if any. """
+        # detach prediction results to cpu
+        pred = {k: v.detach().cpu().numpy() for k, v in pred.items()}
+        # apply score threshold
+        if threshold:
+            selected = pred['scores'] > threshold
+            pred = {k: v[selected] for k, v in pred.items()}
+        if "masks" in pred and mask_threshold:
+            pred["masks"] = pred["masks"] > mask_threshold
+        return pred
+
+    def _get_det_bboxes(
```

Comment:
same with this. now there are two `_get_det_bboxes` methods

Reply:
Removed the old one.

Reply:
Merged with `_process_pred_results()` into `_get_det_bboxes_and_mask()`